tornado/ioloop.py

```diff
@@ -927,6 +897,53 @@ def add_callback_from_signal(self, callback, *args, **kwargs):
                 self._callbacks.append(functools.partial(
                     stack_context.wrap(callback), *args, **kwargs))
 
+    def _handle_waker(self, fd, events):
+        try:
+            self._waker.consume()
+        except:
+            # If consume errors out, we want to reset our waker in case it is
+            # left in a bad state.
+            waker_writer_fileno = self._waker.write_fileno()
+            self.remove_handler(self._waker.fileno())
+            self._waker.close()
```

Comment:
It is unsafe to close the waker while it is still set as the signal wakeup_fd, since the FD could be reassigned in another thread and the signal handler would write garbage to it. The wakeup_fd must be cleared before closing the waker, unless we can separate the code paths so that we only recreate the waker on platforms where the wakeup_fd is not supported.